assets/src/data/model/entity-factory/base-entity.js

```diff
@@ -0,0 +1,203 @@
+import { isArray, upperFirst, camelCase } from 'lodash';
+
+/**
+ * Internal imports
+ */
+import { assertValidSchema } from './assertions';
+import {
+	createGetter,
+	createEntityGettersAndSetters,
+	createPersistingGettersAndSetters,
+	setSaveState,
+} from './create';
+import {
+	SAVE_STATE,
+	PRIVATE_PROPERTIES,
+} from './constants';
+
+/**
+ * BaseEntity is the basic class for all entities.  createEntityFactory returns
+ * an instance of this and all the getters/setters for fields etc are
+ * dynamically created via the constructor.
+ */
+class BaseEntity {
+	[ PRIVATE_PROPERTIES.SAVE_STATE ] = SAVE_STATE.CLEAN;
+	[ PRIVATE_PROPERTIES.VALIDATE_TYPES ] = {};
+
+	/**
+	 * Constructor for Base Entity
+	 * @param {string} modelName
+	 * @param {Object} entityFieldsAndValues
+	 * @param {Object} schema
+	 * @param {Array}fieldPrefixes
+	 * @param {boolean} isNew
+	 */
+	constructor(
+		modelName,
+		entityFieldsAndValues,
+		schema,
+		fieldPrefixes = [],
+		isNew = false,
+	) {
+		assertValidSchema( schema );
+		fieldPrefixes = isArray( fieldPrefixes ) ? fieldPrefixes : [];
+		createGetter( this, 'fieldPrefixes', fieldPrefixes );
+		createGetter( this, 'schema', schema.properties );
+		setSaveState(
+			this,
+			isNew ? SAVE_STATE.NEW : SAVE_STATE.CLEAN
+		);
+		createGetter( this, 'modelName', modelName );
+		createGetter( this, 'originalFieldsAndValues', entityFieldsAndValues );
+		createEntityGettersAndSetters( this );
+		createPersistingGettersAndSetters( this );
+	}
+
+	/**
+	 * Returns the current save state on the entity.
+	 *
+	 * Save state describes whether the entity is:
+	 *
+	 * - SAVE_STATE.NEW: The entity has never been persisted to storage.
+	 * - SAVE_STATE.CLEAN: The entity exists in storage and has not been mutated.
+	 * - SAVE_STATE.DIRTY: The entity is mutated and changes have not been
+	 * persisted to storage.
+	 *
+	 * @return {Symbol}  Returns the current save state for the entity.
+	 */
+	get saveState() {
+		return this[ PRIVATE_PROPERTIES.SAVE_STATE ];
+	}
+
+	/**
+	 * Whether the current save state is SAVE_STATE.NEW
+	 * @return {boolean}  True means SAVE_STATE.NEW is the save state.
+	 */
+	get isNew() {
+		return this.saveState === SAVE_STATE.NEW;
+	}
+
+	/**
+	 * Whether the current save state is SAVE_STATE.DIRTY
+	 * @return {boolean}  True means SAVE_STATE.DIRTY is the save state.
+	 */
+	get isDirty() {
+		return this.saveState === SAVE_STATE.DIRTY;
+	}
+
+	/**
+	 * Whether the current save state is SAVE_STATE.CLEAN
+	 * @return {boolean}  True means SAVE_STATE.CLEAN is the save state.
+	 */
+	get isClean() {
+		return this.saveState === SAVE_STATE.CLEAN;
+	}
+}
+
+/**
+ * A function that gives a class the provided name
+ * (and optionally extends the provided object).
+ * @param {string} name
+ * @param {Object} extendedClass
+ * @return {Function} A function
+ */
+const nameClass = ( name, extendedClass ) => {
+	return class extends extendedClass {
+		static get name() {
+			return name;
+		}
+	};
+};
```

Comment:
This function is used by the model-entity-factory factory for setting the `name` property on the created class.  This allows us to dynamically give the class the name that appears in debugging etc.